### PR TITLE
ZIO Test: Implement Tagged Annotation

### PR DESCRIPTION
**Pull request by adamgfraser**
Allows tagging tests with string tags. Example output:

```
[info]   + monad laws - tagged: "laws", "monad"
[info]     + monad left identity - tagged: "laws", "monad"
[info]     + monad right identity - tagged: "laws", "monad"
[info]     + monad associativity - tagged: "laws", "monad"
```

### CODE REVIEW

File: test/shared/src/main/scala/zio/test/TestAspect.scala [hunk 1]
```diff
@@ -575,6 +575,12 @@ object TestAspect extends TimeoutVariants {
         }
     }
 
+  /**
+   * Annotates tests with string tags.
+   */
+  def tag(tags: String*): TestAspectAtLeastR[Annotations] =
```
Comment (jdegoes) [inline test/shared/src/main/scala/zio/test/TestAspect.scala]:
```suggestion
  def tag(tag: String, tags: String*): TestAspectAtLeastR[Annotations] =
```

**Comment by jdegoes** [comment]
Looks great! Would also be nice to support an option from sbt, like now we have `-t` to run a specific test, maybe we'd have `-tags interruption` to run all tests tagged with `interruption`.

File: test/shared/src/main/scala/zio/test/TestAspect.scala (hunk 1 above)
Comment (adamgfraser) [reply test/shared/src/main/scala/zio/test/TestAspect.scala]: Great suggestion!

**Comment by adamgfraser** [comment]
I think it may be challenging for us to stop a test from running with test annotations. The annotation is added as part of the effectual test (in this case immediately before the original test effect). So by the time we get to the SBT test runner preparing to execute the tests is just has a monolothic test effect where before executing it there are no annotations and after starting to execute it the test is already in motion.

So I see two ways we could approach this. First, we could have the test aspect that adds the annotation immediately check some global state to see if there is a command line argument about tags and if so ignore the test if it doesn’t contain the right tag. However, I’m not sure how we would do this since the test aspect doesn’t have access to anything other than the spec, which doesn’t have this type of configuration information.

The second way would be to build the tags functionality more on the label PR that we have outstanding. We could modify a label to in addition to having a `render` method have a `tags` field. The label exists statically before we start executing any tests so it would be easy for the test runner to look at that and filter out tests based on it like our existing filtering functionality.

**Comment by jdegoes** [comment]
Hmmm. Let me look. I'm interested in what what sort of changes would make it possible to do this using test annotations... e.g. what if every test stored a test annotation map? Then that could be the initial per test map. Could solve labeling problem too.

**Comment by adamgfraser** [comment]
Right now only executed specs have test annotation maps. We could have all specs include them, but I don't actually think that would be sufficient to solve the problem. Modifying the annotation map is an effect and test annotations don't have the ability to perform effects on their own, they can only modify the effects of the tests.

**Comment by jdegoes** [comment]
@adamgfraser 

I was thinking something like this:


```scala
final case class TestCase[-R, +E, +L, +T](
  label: L, test: ZIO[R, E, T], annotations: TestAnnotationMap) extends SpecCase[R, E, L, T, Nothing]
```

where `annotations` is initially set to `TestAnnotationMap.empty`. But aspects could augment it with other "starting" information, such as tags, prior to a run.

In the normal flow, the test runner would run each test with the `annotations` map, and then, replace the value with the new (effectfully computed) `annotations` map following the run.

Could that work?

**Comment by adamgfraser** [comment]
@jdegoes I think so. Let me work on that.

**Comment by adamgfraser** [comment]
@jdegoes This is ready for another review.

File: test/shared/src/main/scala/zio/test/Spec.scala [hunk 2]
```diff
@@ -111,16 +121,46 @@ final case class Spec[-R, +E, +L, +T](caseValue: SpecCase[R, E, L, T, Spec[R, E,
...
-      case TestCase(label, test) =>
-        if (f(label)) Some(Spec.test(label, test)) else None
+      case TestCase(label, test, annotations) =>
+        if (f(label)) Some(Spec.test(label, test, annotations)) else None
+    }
+  }
+
+  /**
+   * Returns a new spec with only those suites and tests with tags satisfying
+   * the specified predicate. If not tags satisfy the specified predicate then
```
Comment (jdegoes) [inline test/shared/src/main/scala/zio/test/Spec.scala]:
```suggestion
   * the specified predicate. If no tags satisfy the specified predicate then
```

File: test/shared/src/main/scala/zio/test/package.scala [hunk 3]
```diff
@@ -110,7 +110,7 @@ package object test extends CompileVariants {
   /**
    * An `ExecutedSpec` is a spec that has been run to produce test results.
    */
-  type ExecutedSpec[+E, +L, +S] = Spec[Any, Nothing, L, Annotated[ExecutedResult[E, S]]]
+  type ExecutedSpec[+E, +L, +S] = Spec[Any, Nothing, L, ExecutedResult[E, S]]
```
Comment (jdegoes) [inline test/shared/src/main/scala/zio/test/package.scala]: This is a nice simplification.

**Pull request merged**